### DISCUSSION TITLE
new neuron model

### DISCUSSION THREAD

**Discussion by tavan95** [norse/norse#361] [General]
I newly start to learn about spiking neural networks. and like to learn about the "norse" platform. I like to know, is it an easy way to define a new neuron model in your platform? or define a new learning rule?

**Comment by Jegp** [comment]
I like to think the answer to that is "yes"! But let me try to convince you with three arguments:

## It's all done in Python
Because we're building on PyTorch, you can write all your equations in Python. But the Python code you write is mapped to hardware accelerators, so it'll run (almost) as fast as if you wrote it in C++/CUDA.

## It's all functions
In PyTorch, you can define operations on tensors as functions. Therefore, implementing a neuron model simply means implementing a function. As an example, writing an [integrate-and-fire function is ~4 lines of code](https://github.com/norse/norse/blob/main/norse/torch/functional/iaf.py#L65) and [STDP is around ~5 lines of code](https://github.com/norse/norse/blob/main/norse/torch/functional/stdp.py#L129) (plus/minus some state handling).

## There's a lot of inspiration
We've already written a lot of neuron models and learning rules you can learn from. Here's a documentation page that explains the architecture of Norse and how to create your own module that---hopefully---will help you out: https://norse.github.io/norse/pages/development.html

Don't hesitate to ping us here if you need any advice. We're also on Discord: https://discord.gg/XzTNwTuw
